LLVM 12 support

Hi, I've been using this tool very happily for a long time while building a compiler. Now, for various reasons, we've upgraded to llvm 12. That breaks this tool.  I don't really have much familiarity with the LLVM implementation or C++ the language, so the errors are hard to interpret. How hard do you think it would be to upgrade this tool to LLVM 12? Any tips/suggestions/...?

Here's what I've tried so far. In `CMakeLists.txt` I've bumped the llvm version to 12. I also had to give an explicit install location for llvm but that's probably my system being weird.

There are some minor compiler errors where LLVM now uses StringRef instead of `std::string`, adding `.str()` fixes those.

Then there is this (DebugIR.cpp, line ~250) which no longer works. 

    DebugLoc NewLoc = DebugLoc::get(Line, Col, Scope, InlinedAt);

Looks like that constructor got removed?  I _think_ that in most cases `NewLoc == Loc`? There doesn't seem to be a similar constructor any more, so not sure what to do with this.

Anyway, that's all still kind of OK, but then I'm really stumped by these linker errors:

```
/usr/bin/ld: CMakeFiles/debugir.dir/DebugIR.cpp.o:(.data.rel.ro+0x10): undefined reference to `typeinfo for llvm::AssemblyAnnotationWriter'
/usr/bin/ld: CMakeFiles/debugir.dir/DebugIR.cpp.o:(.data.rel.ro._ZTIN4llvm18ValueMapCallbackVHIPKNS_5ValueEjNS_14ValueMapConfigIS3_NS_3sys10SmartMutexILb0EEEEEEE[_ZTIN4llvm18ValueMapCallbackVHIPKNS_5ValueEjNS_14ValueMapConfigIS3_NS_3sys10SmartMutexILb0EEEEEEE]+0x10): undefined reference to `typeinfo for llvm::CallbackVH'
/usr/bin/ld: CMakeFiles/debugir.dir/DebugIR.cpp.o:(.data.rel.ro._ZTIN4llvm18ValueMapCallbackVHIPKNS_10BasicBlockEPNS_14DILexicalBlockENS_14ValueMapConfigIS3_NS_3sys10SmartMutexILb0EEEEEEE[_ZTIN4llvm18ValueMapCallbackVHIPKNS_10BasicBlockEPNS_14DILexicalBlockENS_14ValueMapConfigIS3_NS_3sys10SmartMutexILb0EEEEEEE]+0x10): undefined reference to `typeinfo for llvm::CallbackVH'
/usr/bin/ld: CMakeFiles/debugir.dir/DebugIR.cpp.o:(.data.rel.ro._ZTIN4llvm18ValueMapCallbackVHIPKNS_8FunctionEPNS_12DISubprogramENS_14ValueMapConfigIS3_NS_3sys10SmartMutexILb0EEEEEEE[_ZTIN4llvm18ValueMapCallbackVHIPKNS_8FunctionEPNS_12DISubprogramENS_14ValueMapConfigIS3_NS_3sys10SmartMutexILb0EEEEEEE]+0x10): undefined reference to `typeinfo for llvm::CallbackVH'
/usr/bin/ld: CMakeFiles/debugir.dir/DebugIR.cpp.o:(.data.rel.ro._ZTIN4llvm18ValueMapCallbackVHIPKNS_5ValueENS_14WeakTrackingVHENS_14ValueMapConfigIS3_NS_3sys10SmartMutexILb0EEEEEEE[_ZTIN4llvm18ValueMapCallbackVHIPKNS_5ValueENS_14WeakTrackingVHENS_14ValueMapConfigIS3_NS_3sys10SmartMutexILb0EEEEEEE]+0x10): undefined reference to `typeinfo for llvm::CallbackVH'
/usr/bin/ld: CMakeFiles/debugir.dir/Main.cpp.o:(.data.rel.ro._ZTIN4llvm2cl15OptionValueCopyINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE[_ZTIN4llvm2cl15OptionValueCopyINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEEE]+0x10): undefined reference to `typeinfo for llvm::cl::GenericOptionValue'
/usr/bin/ld: CMakeFiles/debugir.dir/Main.cpp.o:(.data.rel.ro._ZTIN4llvm2cl15OptionValueCopyIbEE[_ZTIN4llvm2cl15OptionValueCopyIbEE]+0x10): undefined reference to `typeinfo for llvm::cl::GenericOptionValue'
collect2: error: ld returned 1 exit status
```

maybe relevant detail: my llvm "installation" is an archive downloaded from the llvm github. The install script does not seem to work on my version of ubuntu. Crucially that archive does not seem to contain the `.cpp` files, so maybe that's the problem?

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -13,11 +13,15 @@ message(STATUS "Build type: ${CMAKE_BUILD_TYPE}")
 # check dependencies
 #
 # LLVM stuff
-find_package(LLVM 10 REQUIRED CONFIG)
+find_package(LLVM 12 REQUIRED CONFIG)
 message(STATUS "Found LLVM ${LLVM_PACKAGE_VERSION}")
 message(STATUS "Using LLVMConfig.cmake in: ${LLVM_DIR}")
 # Find the libraries that correspond to the LLVM components we use
 llvm_map_components_to_libnames(llvm_libs support core irreader transformutils)
+if (NOT LLVM_ENABLE_RTTI)
+  message(STATUS "LLVM does not have RTTI enabled. Building with -fno-rtti")
+  add_compile_options(-fno-rtti)
+endif()
 
 # compiler and linker options
 add_compile_options(-Wall)

--- a/DebugIR.cpp
+++ b/DebugIR.cpp
@@ -296,7 +296,6 @@ private:
   }
 
   DIScope *getBlockScope(DIScope *ParentScope, const BasicBlock *B) {
-    const Function *F = B->getParent();
     auto BScope = BlockDescriptors.find(B);
     if (BScope != BlockDescriptors.end()) {
       return BScope->second;
@@ -306,7 +305,7 @@ private:
       if (!findLine(B, Line)) {
         LLVM_DEBUG(dbgs() << "WARNING: No line for basic block "
                           << B->getName().str() << " in Function "
-                          << F->getName().str() << "\n");
+                          << B->getParent()->getName().str() << "\n");
       }
       auto Scope = Builder.createLexicalBlock(ParentScope, FileNode, Line, 0);
       BlockDescriptors[B] = Scope;

--- a/DebugIR.cpp
+++ b/DebugIR.cpp
@@ -239,12 +239,12 @@ public:
       return;
     }
 
-    DIScope *Scope;
+    DILocalScope *Scope;
     DILocation *InlinedAt;
     if (Loc) {
-      Scope = llvm::cast<DIScope>(Loc.getScope());
+      Scope = llvm::cast<DILocalScope>(Loc.getScope());
       InlinedAt = Loc.getInlinedAt();
-    } else if ((Scope = findScope(&I))) {
+    } else if ((Scope = dyn_cast<DILocalScope>(findScope(&I)))) {
       InlinedAt = nullptr;
     } else {
       LLVM_DEBUG(dbgs() << "WARNING: no valid scope for instruction " << &I
@@ -253,7 +253,8 @@ public:
       return;
     }
 
-    DebugLoc NewLoc = DebugLoc::get(Line, Col, Scope, InlinedAt);
+    DebugLoc NewLoc =
+        DebugLoc(DILocation::get(M.getContext(), Line, Col, Scope, InlinedAt));
     addDebugLocation(I, NewLoc);
 
     if (!I.getType()->isVoidTy() && !I.getName().empty()) {
@@ -276,7 +277,7 @@ private:
       // save fields from existing CU to re-use in the new CU
       Producer = CUToReplace->getProducer();
       IsOptimized = CUToReplace->isOptimized();
-      Flags = CUToReplace->getFlags();
+      Flags = CUToReplace->getFlags().str();
       RuntimeVersion = CUToReplace->getRuntimeVersion();
       SplitName = CUToReplace->getSplitDebugFilename();
     } else {

--- a/Main.cpp
+++ b/Main.cpp
@@ -74,7 +74,7 @@ int main(int argc, char *argv[]) {
     PM.run(*M);
   }
 
-  auto DisplayM = createDebugInfo(*M.get(), Directory, Filename);
+  auto DisplayM = createDebugInfo(*M.get(), Directory.str(), Filename.str());
   std::error_code EC;
 
   // Update InputFile in-place so that line numbers match with DebugFile.

--- a/README.md
+++ b/README.md
@@ -30,11 +30,14 @@ Such a file can be dynamically generated, or using clang as
 </details>
 
 ### Clone and build
-This tool requires LLVM-10 to be installed.
+This tool requires LLVM-12 to be installed.
   - `$git clone https://github.com/vaivaswatha/debugir.git debugir`
   - `$cd debugir; mkdir build; cd build`
   - `$cmake -DCMAKE_BUILD_TYPE=Release ../`
   - `$cmake --build .`
+
+If you have LLVM installed in a non-standard path, you may provide the
+additional `CMake` argument `-DLLVM_DIR=/path/to/llvm`.
 
 ### Run
 You should now have an executable file `buildir` in your CMake build
@@ -57,11 +60,13 @@ the `-instnamer` flag to `debugir`.
 Following on the example [here](https://llvm.org/docs/DebuggingJITedCode.html)
 let us try and debug `hello.dbg.ll`.
 
-  - `$gdb lli-10`
+  - `$gdb lli`
   - `(gdb) set args -jit-kind=mcjit hello.dbg.ll`
   - `(gdb) break hello.ll:25 # set breakpoint at line 25 in hello.ll`
   - `(gdb) run`
 
 You should now hit the program at line 25 in `hello.ll`, assuming that
 line 25 is a valid line number in the LLVM source. Change this line number
-to an appropriate value or to a function name.
+to an appropriate value or to a function name. Note: Since `lli`, at the
+time of invocation from gdb will not have, yet, loaded the object file for
+`hello`, you will need to set `set breakpoint pending on` in `gdb`.

